File: server/mistralAnalyzer.js
// mistralAnalyzer.js - Analyseur de bouteilles de vin avec Google Vision + Mistral AI
// Version 1.2 - NOUVEAU WORKFLOW : Google Vision (nom + année) → Mistral (complétion des champs)

const mistralConfig = require("./mistralConfig");
const googleVision = require("./googleVision");
const mistralAI = require("./mistralAI");

// Importer callMistral depuis mistralAI
const { callMistral } = mistralAI;

// Vérifier que callMistral est bien défini
if (typeof callMistral !== 'function') {
    console.error('[MistralAnalyzer] callMistral is not defined!');
    console.error('[MistralAnalyzer] Available functions in mistralAI:', Object.keys(mistralAI));
}

/**
 * Retourne des informations par défaut pour une bouteille en cas d'erreur
 * @param {string} errorMessage - Message d'erreur
 * @returns {Object} - Objet bouteille avec valeurs null
 */
function getFallbackBottleInfo(errorMessage) {
    console.warn(`[MistralAnalyzer] Fallback: ${errorMessage}`);
    return {
        name: null,
        year: null,
        grapes: null,
        region: null,
        appellation: null,
        producer: null,
        country: null,
        alcohol: null,
        drinkFrom: null,
        drinkTo: null,
        foodPairing: null,
        temperature: null,
        photo: null,
        analysisMethod: "Fallback (erreur)",
        extractedText: null,
        error: errorMessage || "Erreur inconnue"
    };
}

/**
 * Complète les informations de la bouteille avec des valeurs par défaut
 * @param {Object} bottleInfo - Informations de la bouteille depuis Mistral
 * @returns {Object} - Informations complètes
 */
function completeBottleInfo(bottleInfo) {
    if (!bottleInfo) {
        return getFallbackBottleInfo("Aucune information de bouteille");
    }
    
    return {
        name: bottleInfo.name || null,
        year: bottleInfo.year ? parseInt(bottleInfo.year) : null,
        grapes: bottleInfo.grapes || null,
        region: bottleInfo.region || bottleInfo.appellation || null,
        appellation: bottleInfo.appellation || null,
        producer: bottleInfo.producer || null,
        country: bottleInfo.country || null,
        alcohol: bottleInfo.alcohol ? parseFloat(bottleInfo.alcohol) : null,
        drinkFrom: bottleInfo.drinkFrom ? parseInt(bottleInfo.drinkFrom) : null,
        drinkTo: bottleInfo.drinkTo ? parseInt(bottleInfo.drinkTo) : null,
        foodPairing: bottleInfo.foodPairing || null,
        temperature: bottleInfo.temperature || null
    };
}

/**
 * Analyse du texte avec Mistral AI pour compléter les informations de la bouteille
 * NOUVEAU : Reçoit le nom et l'année, et complète les autres champs
 * @param {string} name - Nom de la bouteille
 * @param {number|null} year - Année de la bouteille
 * @returns {Promise<Object|null>} - Résultat de l'analyse
 */
async function analyzeBottleWithMistral(name, year) {
    if (!name || !name.trim()) {
        return null;
    }

    // NOUVEAU PROMPT : On fournit le nom et l'année, Mistral complète les autres champs
    const prompt = `Tu es un expert en vin. J'ai une bouteille avec les informations suivantes :
    - Nom : "${name}"
    ${year ? `- Année : ${year}` : ''}

    Complète TOUTES les informations manquantes pour cette bouteille avec tes connaissances sur les vins.
    Retourne UNIQUEMENT un objet JSON valide avec TOUS ces champs (même si null) :
    {
        "name": "${name}",
        "year": ${year || 'null'},
        "grapes": "Cépage(s) principal(aux) (ex: Cabernet Sauvignon, Merlot)",
        "region": "Région ou appellation (ex: Bordeaux, Bourgogne)",
        "appellation": "Appellation officielle (ex: Pauillac, Saint-Émilion)",
        "producer": "Producteur ou domaine (ex: Château Margaux)",
        "country": "Pays (ex: France, Italie)",
        "alcohol": 13.5, // Degré d'alcool (nombre ou null)
        "drinkFrom": 2025, // Année de début de consommation optimale (nombre ou null)
        "drinkTo": 2035, // Année de fin de consommation optimale (nombre ou null)
        "foodPairing": "Accords mets-vins (ex: Viandes rouges, Fromages)",
        "temperature": "Température de service (ex: 16-18°C)"
    }
    Règles strictes :
    - Si une information n'est pas connue, utilise null.
    - Les champs "year", "alcohol", "drinkFrom", "drinkTo" doivent être des nombres ou null.
    - Pour drinkFrom et drinkTo : estime la période optimale de consommation en fonction de l'année et du type de vin.
    - Pour foodPairing : propose des accords mets-vins typiques pour ce type de vin.
    - Pour temperature : indique la température de service idéale.
    - Ne réponds JAMAIS autre chose que le JSON.`;

    try {
        const response = await callMistral(prompt, mistralConfig.analysisSystemPrompt);

        // Nettoyer la réponse
        let cleanedResponse = response.trim();
        cleanedResponse = cleanedResponse.replace(/^```json\s*/, '');
        cleanedResponse = cleanedResponse.replace(/```\s*$/, '');
        cleanedResponse = cleanedResponse.replace(/^```\s*/, '');

        // Extraire le JSON
        const jsonMatch = cleanedResponse.match(/\{[\s\S]*\}/);
        if (jsonMatch) {
            try {
                const result = JSON.parse(jsonMatch[0]);
                // Forcer tous les champs à exister et nettoyer les valeurs
                return {
                    name: result.name || name || null,
                    year: result.year !== undefined ? (result.year === null ? null : parseInt(result.year)) : (year || null),
                    grapes: result.grapes || null,
                    region: result.region || null,
                    appellation: result.appellation || null,
                    producer: result.producer || null,
                    country: result.country || null,
                    alcohol: result.alcohol !== undefined ? (result.alcohol === null ? null : parseFloat(result.alcohol)) : null,
                    drinkFrom: result.drinkFrom !== undefined ? (result.drinkFrom === null ? null : parseInt(result.drinkFrom)) : null,
                    drinkTo: result.drinkTo !== undefined ? (result.drinkTo === null ? null : parseInt(result.drinkTo)) : null,
                    foodPairing: result.foodPairing || null,
                    temperature: result.temperature || null
                };
            } catch (e) {
                console.error("Erreur de parsing JSON :", e);
                return null;
            }
        }
        return null;
    } catch (error) {
        console.error("Erreur Mistral :", error.message);
        return null;
    }
}

/**
 * Analyse du texte avec Mistral AI (ancienne version pour compatibilité)
 * @param {string} text - Texte à analyser
 * @returns {Promise<Object|null>} - Résultat de l'analyse
 */
async function analyzeLabelTextWithMistral(text) {
    if (!text || !text.trim()) {
        return null;
    }

    // Ancien prompt pour compatibilité
    const prompt = `Tu es un expert en vin. Analyse ce texte d'étiquette de vin et extrais TOUTES les informations possibles.
    Texte de l'étiquette : """${text}"""

    Retourne UNIQUEMENT un objet JSON valide avec TOUS ces champs (même si null) :
    {
        "name": "Nom du vin ou du domaine (ex: Château Margaux)",
        "year": 2020, // Année (nombre entier ou null)
        "grapes": "Cépage(s) principal(aux) (ex: Cabernet Sauvignon, Merlot)",
        "region": "Région ou appellation (ex: Bordeaux, Bourgogne)",
        "appellation": "Appellation officielle (ex: Pauillac, Saint-Émilion)",
        "producer": "Producteur ou domaine (ex: Château Margaux)",
        "country": "Pays (ex: France, Italie)",
        "alcohol": 13.5, // Degré d'alcool (nombre ou null)
        "drinkFrom": 2025, // Année de début de consommation (nombre ou null)
        "drinkTo": 2035, // Année de fin de consommation (nombre ou null)
        "foodPairing": "Accords mets-vins (ex: Viandes rouges, Fromages)",
        "temperature": "Température de service (ex: 16-18°C)"
    }
    Règles :
    - Si une information n'est pas présente dans le texte, utilise null.
    - Les champs "year", "alcohol", "drinkFrom", "drinkTo" doivent être des nombres ou null.
    - Ne réponds JAMAIS autre chose que le JSON.`;

    try {
        const response = await callMistral(prompt, mistralConfig.analysisSystemPrompt);

        // Nettoyer la réponse
        let cleanedResponse = response.trim();
        cleanedResponse = cleanedResponse.replace(/^```json\s*/, '');
        cleanedResponse = cleanedResponse.replace(/```\s*$/, '');
        cleanedResponse = cleanedResponse.replace(/^```\s*/, '');

        // Extraire le JSON
        const jsonMatch = cleanedResponse.match(/\{[\s\S]*\}/);
        if (jsonMatch) {
            try {
                const result = JSON.parse(jsonMatch[0]);
                return {
                    name: result.name || null,
                    year: result.year !== undefined ? (result.year === null ? null : parseInt(result.year)) : null,
                    grapes: result.grapes || null,
                    region: result.region || null,
                    appellation: result.appellation || null,
                    producer: result.producer || null,
                    country: result.country || null,
                    alcohol: result.alcohol !== undefined ? (result.alcohol === null ? null : parseFloat(result.alcohol)) : null,
                    drinkFrom: result.drinkFrom !== undefined ? (result.drinkFrom === null ? null : parseInt(result.drinkFrom)) : null,
                    drinkTo: result.drinkTo !== undefined ? (result.drinkTo === null ? null : parseInt(result.drinkTo)) : null,
                    foodPairing: result.foodPairing || null,
                    temperature: result.temperature || null
                };
            } catch (e) {
                console.error("Erreur de parsing JSON :", e);
                return null;
            }
        }
        return null;
    } catch (error) {
        console.error("Erreur Mistral :", error.message);
        return null;
    }
}

/**
 * Analyse une bouteille avec texte manuel (sans image)
 * @param {Object} manualData - Données manuelles {name, year, grapes, region}
 * @returns {Promise<Object>} - Informations de la bouteille
 */
async function analyzeWithManualText(manualData) {
    const { name, year, grapes, region } = manualData;
    const text = `${name || ''} ${year || ''} ${grapes || ''} ${region || ''}`;
    
    if (!text.trim()) {
        return getFallbackBottleInfo("Aucun texte à analyser");
    }
    
    try {
        const bottleInfo = await analyzeLabelTextWithMistral(text);
        if (!bottleInfo) {
            return getFallbackBottleInfo("Aucune information extraite du texte");
        }
        return completeBottleInfo(bottleInfo);
    } catch (error) {
        return getFallbackBottleInfo(error.message);
    }
}

/**
 * NOUVEAU WORKFLOW : Analyse une bouteille avec Google Vision + Mistral AI
 * Étape 1 : Google Vision extrait le nom et l'année
 * Étape 2 : Mistral complète les autres champs (cépage, période, accords, température)
 * @param {string|null} imagePathOrBase64 - Chemin de l'image ou base64
 * @param {boolean} isBase64 - Si true, imagePathOrBase64 est en base64
 * @param {Object|null} manualData - Données manuelles {name, year} si Google Vision échoue
 * @returns {Promise<Object>} - Résultat complet de l'analyse
 */
async function analyzeBottleWithTwoStepProcess(imagePathOrBase64, isBase64 = false, manualData = null) {
    if (!mistralConfig.apiKey) {
        return {
            ...getFallbackBottleInfo("Mistral AI non configurée - Ajoutez MISTRAL_API_KEY dans .env"),
            requiresManualInput: false,
            googleVisionAvailable: googleVision.isGoogleVisionConfigured()
        };
    }
    
    const googleVisionAvailable = googleVision.isGoogleVisionConfigured();
    let name = null;
    let year = null;
    let rawText = null;
    let analysisMethod = "";
    
    try {
        // ÉTAPE 1 : Extraire le nom et l'année avec Google Vision
        if (imagePathOrBase64 && googleVisionAvailable) {
            try {
                const visionResult = await googleVision.analyzeBottleForNameAndYear(imagePathOrBase64, isBase64);
                if (visionResult.success) {
                    name = visionResult.name;
                    year = visionResult.year;
                    rawText = visionResult.rawText;
                    analysisMethod = "Google Vision + Mistral AI";
                } else {
                    console.warn("Google Vision n'a pas pu extraire le nom/année :", visionResult.error);
                    analysisMethod = "Mistral AI (sans OCR)";
                }
            } catch (visionError) {
                console.warn("Erreur Google Vision :", visionError.message);
                analysisMethod = "Mistral AI (sans OCR)";
            }
        } else if (!googleVisionAvailable) {
            analysisMethod = "Mistral AI (sans OCR)";
        }
        
        // Si on a des données manuelles, les utiliser
        if (manualData) {
            if (manualData.name) {
                name = manualData.name;
            }
            // TOUJOURS utiliser l'année manuelle si fournie (même si name est vide)
            if (manualData.year !== undefined && manualData.year !== null && manualData.year !== '') {
                year = parseInt(manualData.year);
            }
            if (analysisMethod) {
                analysisMethod += " + Saisie manuelle";
            } else {
                analysisMethod = "Saisie manuelle + Mistral AI";
            }
        }
        
        // ÉTAPE 2 : Compléter avec Mistral AI
        let bottleInfo = null;
        if (name) {
            bottleInfo = await analyzeBottleWithMistral(name, year);
        } else if (rawText) {
            // Fallback : analyser le texte brut avec l'ancien système
            bottleInfo = await analyzeLabelTextWithMistral(rawText);
        }
        
        // Si on n'a toujours pas d'info, essayer avec un prompt générique
        if (!bottleInfo && imagePathOrBase64) {
            console.log("Tentative d'analyse avec Mistral uniquement (sans OCR)");
            const genericText = "Analyse cette bouteille de vin et extrais toutes les informations possibles : nom, année, cépage, région, appellation, producteur, pays, degré d'alcool, période de consommation, accords mets-vins, température de service.";
            bottleInfo = await analyzeLabelTextWithMistral(genericText);
        }
        
        if (!bottleInfo) {
            return {
                ...getFallbackBottleInfo("Aucune information extraite"),
                requiresManualInput: !name || !year,
                missingFields: {
                    name: !name,
                    year: !year
                },
                partialData: { name, year },
                analysisMethod,
                googleVisionAvailable
            };
        }
        
        const completeInfo = completeBottleInfo(bottleInfo);
        
        // Vérifier si l'année est manquante dans les résultats finaux
        const finalYear = completeInfo.year;
        const finalName = completeInfo.name;
        const hasMissingFields = !finalName || !finalYear;
        
        return {
            ...completeInfo,
            analysisMethod,
            extractedText: rawText,
            mistralAvailable: true,
            googleVisionAvailable,
            requiresManualInput: hasMissingFields,
            missingFields: hasMissingFields ? {
                name: !finalName,
                year: !finalYear
            } : null,
            partialData: hasMissingFields ? { name: finalName, year: finalYear } : null
        };
    } catch (error) {
        return {
            ...getFallbackBottleInfo(`Erreur: ${error.message}`),
            requiresManualInput: !name || !year,
            missingFields: {
                name: !name,
                year: !year
            },
            partialData: { name, year },
            analysisMethod,
            googleVisionAvailable
        };
    }
}

/**
 * Analyse une bouteille avec Google Vision + Mistral AI (ancienne version pour compatibilité)
 * @param {string|null} imagePathOrBase64 - Chemin de l'image ou base64
 * @param {boolean} isBase64 - Si true, imagePathOrBase64 est en base64
 * @returns {Promise<Object>} - Résultat complet de l'analyse
 */
async function analyzeBottleWithMistralOnly(imagePathOrBase64, isBase64 = false) {
    if (!mistralConfig.apiKey) {
        return getFallbackBottleInfo("Mistral AI non configurée - Ajoutez MISTRAL_API_KEY dans .env");
    }
    
    const googleVisionAvailable = googleVision.isGoogleVisionConfigured();
    
    try {
        let labelText = null;
        if (imagePathOrBase64) {
            if (googleVisionAvailable) {
                try {
                    labelText = await googleVision.getCleanedLabelText(imagePathOrBase64, isBase64);
                } catch (visionError) {
                    console.warn("OCR échoué, tentative avec Mistral uniquement :", visionError.message);
                    labelText = "Image de bouteille de vin à analyser";
                }
            } else {
                console.warn("Google Vision non configuré, analyse avec Mistral uniquement");
                labelText = "Analyse cette image de bouteille de vin et extrais toutes les informations possibles.";
            }
        }
        
        let bottleInfo = null;
        if (labelText && labelText.trim()) {
            bottleInfo = await analyzeLabelTextWithMistral(labelText);
        }
        
        if (!bottleInfo && imagePathOrBase64) {
            console.log("Tentative d'analyse avec Mistral uniquement (sans OCR)");
            const genericText = "Analyse cette bouteille de vin et extrais toutes les informations possibles : nom, année, cépage, région, appellation, producteur, pays, degré d'alcool, période de consommation, accords mets-vins, température de service.";
            bottleInfo = await analyzeLabelTextWithMistral(genericText);
        }
        
        if (!bottleInfo) {
            return getFallbackBottleInfo("Aucune information extraite");
        }
        
        const completeInfo = completeBottleInfo(bottleInfo);
        return {
            ...completeInfo,
            analysisMethod: labelText ? (googleVisionAvailable ? "Google Vision + Mistral AI" : "Mistral AI (sans OCR)") : "Mistral AI (analyse générique)",
            extractedText: labelText,
            mistralAvailable: true,
            googleVisionAvailable: googleVisionAvailable
        };
    } catch (error) {
        return getFallbackBottleInfo(`Erreur: ${error.message}`);
    }
}

module.exports = {
    analyzeBottleWithMistralOnly,
    analyzeBottleWithTwoStepProcess,
    getFallbackBottleInfo,
    analyzeLabelTextWithMistral,
    analyzeBottleWithMistral,
    completeBottleInfo,
    analyzeWithManualText
};

File: server_flask/mistral_analyzer.py
"""
Analyseur de bouteilles de vin avec Mistral AI et Google Vision
"""
import json
import re


class MistralAnalyzer:
    def __init__(self, mistral_ai_instance, google_vision_instance, config):
        self.mistral = mistral_ai_instance
        self.google_vision = google_vision_instance
        self.config = config
        
        # Prompts système
        self.SYSTEM_PROMPT = """Tu es un expert en vin et en gestion de cave à vin. 
Tu dois répondre de manière précise, professionnelle et utile aux questions sur :
- Les accords mets-vins (quel vin avec quel plat)
- Les températures de service
- Les cépages, régions et appellations
- La gestion d'une cave à vin
- L'analyse d'étiquettes de vin

Réponds toujours en français, de manière claire et concise. 
Si tu ne connais pas la réponse, dis-le honnêtement et propose des alternatives.
Ne fais pas de blagues, reste professionnel.
Utilise des emojis vinicoles (🍷, 🍇) avec modération."""
        
        self.ANALYSIS_SYSTEM_PROMPT = """Tu es un expert en reconnaissance d'étiquettes de vin. 
On va te donner du texte extrait d'une étiquette de vin à analyser.
Ton rôle est d'analyser ce texte et d'extraire les informations suivantes :
- Nom du vin (ou du domaine/château)
- Année/millésime (si présente)
- Cépage(s) principal(aux)
- Région/appellation
- Producteur (si identifiable)
- Pays d'origine
- Degré d'alcool

Format de réponse : UNIQUEMENT un objet JSON avec les champs : name, year, grapes, region, appellation, producer, country, alcohol.
Si une information n'est pas trouvée, mets null.
Ne réponds JAMAIS autre chose que le JSON."""
    
    def get_fallback_bottle_info(self, error_message=None):
        """Retourner des informations par défaut pour une bouteille"""
        return {
            'name': None,
            'year': None,
            'grapes': None,
            'region': None,
            'appellation': None,
            'producer': None,
            'country': None,
            'alcohol': None,
            'drinkFrom': None,
            'drinkTo': None,
            'foodPairing': None,
            'temperature': None,
            'analysisMethod': 'Fallback',
            'extractedText': error_message or 'Aucune information disponible',
            'requiresManualInput': True,
            'missingFields': {'name': True, 'year': True}
        }
    
    def clean_extracted_text(self, text):
        """Nettoyer le texte extrait"""
        if not text:
            return ''
        
        # Supprimer les caractères non imprimables
        text = ''.join(char for char in text if char.isprintable() or char.isspace())
        
        # Remplacer les espaces multiples par un seul espace
        text = re.sub(r'\s+', ' ', text)
        
        # Supprimer les espaces au début et à la fin
        text = text.strip()
        
        return text
    
    def analyze_bottle_with_mistral_only(self, image_path=None, is_base64=False):
        """Analyser une bouteille avec Mistral AI uniquement (sans OCR)"""
        try:
            # Si c'est une image base64, essayer de l'analyser directement
            if is_base64 and image_path:
                # Pour l'instant, on ne peut pas analyser directement l'image avec Mistral
                # On retourne des informations par défaut
                return self.get_fallback_bottle_info('Analyse directe non disponible')
            
            # Sinon, retourner des informations par défaut
            return self.get_fallback_bottle_info('Aucune image fournie')
            
        except Exception as e:
            print(f"Erreur lors de l'analyse avec Mistral uniquement: {e}")
            return self.get_fallback_bottle_info(str(e))
    
    def analyze_bottle_with_two_step_process(self, image_path=None, is_base64=False, manual_data=None, correct_name=True):
        """Analyser une bouteille en deux étapes : OCR puis Mistral"""
        try:
            extracted_text = None
            
            # Étape 1 : Extraire le texte avec Google Vision
            if image_path:
                if is_base64:
                    extracted_text = self.google_vision.extract_text_from_base64(image_path)
                else:
                    extracted_text = self.google_vision.extract_text_from_image(image_path)
            
            # Nettoyer le texte extrait
            if extracted_text:
                extracted_text = self.clean_extracted_text(extracted_text)
            
            # Étape 2 : Analyser avec Mistral
            if extracted_text:
                analysis = self.mistral.analyze_bottle_text(extracted_text)
                if analysis:
                    # Fusionner avec les données manuelles si fournies
                    if manual_data:
                        for key, value in manual_data.items():
                            if value and key in analysis:
                                analysis[key] = value
                    
                    # Formater le résultat
                    result = {
                        'name': analysis.get('name'),
                        'year': analysis.get('year'),
                        'grapes': analysis.get('grapes'),
                        'region': analysis.get('region') or analysis.get('appellation'),
                        'appellation': analysis.get('appellation'),
                        'producer': analysis.get('producer'),
                        'country': analysis.get('country'),
                        'alcohol': analysis.get('alcohol'),
                        'drinkFrom': None,
                        'drinkTo': None,
                        'foodPairing': None,
                        'temperature': None,
                        'analysisMethod': 'Google Vision + Mistral',
                        'extractedText': extracted_text,
                        'requiresManualInput': False,
                        'missingFields': None,
                        'partialData': None
                    }
                    
                    # Déterminer les champs manquants
                    missing_fields = {}
                    if not result['name']:
                        missing_fields['name'] = True
                    if not result['year']:
                        missing_fields['year'] = True
                    
                    result['requiresManualInput'] = bool(missing_fields)
                    result['missingFields'] = missing_fields if missing_fields else None
                    
                    return result
            
            # Si on arrive ici, l'analyse a échoué
            return self.get_fallback_bottle_info('Échec de l\'analyse OCR')
            
        except Exception as e:
            print(f"Erreur lors de l'analyse en deux étapes: {e}")
            return self.get_fallback_bottle_info(str(e))
    
    def analyze_with_manual_text(self, manual_data):
        """Analyser avec du texte manuel"""
        try:
            # Construire un texte à partir des données manuelles
            text_parts = []
            if manual_data.get('name'):
                text_parts.append(f"Nom: {manual_data['name']}")
            if manual_data.get('year'):
                text_parts.append(f"Année: {manual_data['year']}")
            if manual_data.get('grapes'):
                text_parts.append(f"Cépage: {manual_data['grapes']}")
            if manual_data.get('region'):
                text_parts.append(f"Région: {manual_data['region']}")
            
            if not text_parts:
                return self.get_fallback_bottle_info('Aucune donnée manuelle fournie')
            
            extracted_text = "\n".join(text_parts)
            
            # Analyser avec Mistral
            analysis = self.mistral.analyze_bottle_text(extracted_text)
            
            if analysis:
                # Fusionner avec les données manuelles
                for key, value in manual_data.items():
                    if value and key in analysis:
                        analysis[key] = value
                
                result = {
                    'name': analysis.get('name') or manual_data.get('name'),
                    'year': analysis.get('year') or manual_data.get('year'),
                    'grapes': analysis.get('grapes') or manual_data.get('grapes'),
                    'region': analysis.get('region') or analysis.get('appellation') or manual_data.get('region'),
                    'appellation': analysis.get('appellation'),
                    'producer': analysis.get('producer'),
                    'country': analysis.get('country'),
                    'alcohol': analysis.get('alcohol'),
                    'drinkFrom': None,
                    'drinkTo': None,
                    'foodPairing': None,
                    'temperature': None,
                    'analysisMethod': 'Analyse texte manuel',
                    'extractedText': extracted_text
                }
                return result
            
            # Retourner les données manuelles si l'analyse échoue
            return {
                'name': manual_data.get('name'),
                'year': manual_data.get('year'),
                'grapes': manual_data.get('grapes'),
                'region': manual_data.get('region'),
                'analysisMethod': 'Données manuelles',
                'extractedText': extracted_text
            }
            
        except Exception as e:
            print(f"Erreur lors de l'analyse avec texte manuel: {e}")
            return self.get_fallback_bottle_info(str(e))
    
    def generate_fallback_response_with_cave(self, question, bottles):
        """Générer une réponse de secours avec le contexte de la cave"""
        if not bottles:
            return "Je n'ai pas accès à votre cave pour répondre à cette question."
        
        # Analyser la question pour comprendre l'intention
        question_lower = question.lower()
        
        # Si la question concerne des recommandations
        if any(word in question_lower for word in ['recommande', 'conseil', 'quel vin', 'suggère', 'propose']):
            # Filtrer les bouteilles prêtes à boire
            ready_to_drink = [b for b in bottles 
                           if b.get('drinkFrom') and b.get('drinkTo')
                           and b['drinkFrom'] <= 2026 <= b['drinkTo']]
            
            if ready_to_drink:
                return f"Je vous recommande l'un de ces vins de votre cave qui sont prêts à boire : {', '.join(b['name'] for b in ready_to_drink[:3])}"
            else:
                return f"Dans votre cave, vous avez ces vins : {', '.join(b['name'] for b in bottles[:5])}. Lequel souhaitez-vous déguster ?"
        
        # Si la question concerne un type de vin spécifique
        if any(word in question_lower for word in ['rouge', 'blanc', 'rosé', 'champagne']):
            wine_type = 'rouge' if 'rouge' in question_lower else 'blanc' if 'blanc' in question_lower else 'rosé' if 'rosé' in question_lower else 'champagne'
            matching = [b for b in bottles if wine_type in (b.get('name', '') + b.get('grapes', '')).lower()]
            if matching:
                return f"Dans votre cave, vous avez ces vins {wine_type}s : {', '.join(b['name'] for b in matching)}"
        
        # Réponse générique
        return f"Dans votre cave, vous avez {len(bottles)} bouteilles. Je peux vous aider à choisir un vin si vous me donnez plus de détails."
